src/pip/_internal/operations/install/wheel.py

```diff
@@ -595,9 +595,11 @@ def pyc_source_file_paths() -> Generator[str, None, None]:
         # Sorting installation paths makes it easier to reproduce and debug
         # issues related to permissions on existing files.
         for installed_path in sorted(set(installed.values())):
-            full_installed_path = os.path.join(lib_dir, installed_path)
+            full_installed_path = os.path.abspath(os.path.join(lib_dir, installed_path))
             if not os.path.isfile(full_installed_path):
                 continue
+            if lib_dir != os.path.commonpath([lib_dir, full_installed_path]):
+                continue
```

Comment:
It seems that this will only compile files installed into the "main" library directory (either platlib or purelib), but I think it would make sense to compile files installed into both.

Reply:
A wheel can only install Python source files to _either_ platlib or purelib, and can only access the other via the [`.data`](https://packaging.python.org/en/latest/specifications/binary-distribution-format/#the-data-directory) mechanism. Those files should be considered data files—you can still install Python source files there, sure, but pip should treat them as data files.

Reply:
@uranusjr I think this is contradicted by the [wheel spec](https://packaging.python.org/en/latest/specifications/binary-distribution-format/#what-s-the-deal-with-purelib-vs-platlib):

> A wheel with “Root-Is-Purelib: false” with all its files in `{name}-{version}.data/purelib` is equivalent to a wheel with “Root-Is-Purelib: true” with those same files in the root, and it is legal to have files in both the “purelib” and “platlib” categories.

If these wheels are equivalent, then pip should treat `.py` files the same regardless of whether they are in the root of the wheel or in `{name}-{version}.data/purelib` (or `platlib`).

Reply:
The spec also says the installer should *Compile any installed .py to .pyc*; if you want to follow the spec to the letter, custom scripts _should_ be compiled. So… make your pick 🙂

Reply:
FWIW, I was able to find several popular projects whose wheels install `.py` files via `{name}-{version}.data/purelib` / `platlib`:

* libclang
* pymediainfo
* pymongocrypt
* semgrep
* tensorflow-addons
* tritonclient

With this patch (and assuming purelib and platlib are different directories), those `.py` files will no longer be compiled.